### DISCUSSION TITLE
Combination of the theories of Reals and FloatingPoint

### DISCUSSION THREAD

**Discussion by nicola-gigante** [cvc5/cvc5#8024] [Q&A] [answered]
Hi all,

I've been playing with CVC5 to do some verification of floating point code and it seems that it can handle mixed Real and FloatingPoint constraints very well. 

For example:

```
(set-logic ALL)

(define-const onef Float32 ((_ to_fp 8 24) RNE 1))
(declare-const xf Float32)
(declare-const yf Float32)

(declare-const xr Real)
(declare-const yr Real)

(define-fun fabs ((x Real)) Real (ite (< x 0) (- x) x))

(assert (< (+ xr yr) 1))
(assert (fp.lt (fp.add RNE xf yf) onef))
(assert (<= (fabs (- xr (fp.to_real xf))) 0.00002))
(assert (<= (fabs (- yr (fp.to_real yf))) 0.00008))
(assert (>= xr 0))
(assert (<= xr 100))
(assert (>= yr 0))
(assert (<= yr 100))

(check-sat)
(get-model)
```

So I'm are wondering what is happening under the hood. In particular, what are the theoretical guarantees for the theory combination of reals and floating points? In principle, the combined quantifier-free fragment can be undecidable, so I'm wondering whether some well-known transfer result for satisfiability (such as Nelson-Oppen or Polite Combination) is employed. Floating Points do not seem to enjoy stable infiniteness, so I doubt that the Nelson-Oppen method can be applied.

Thank you very much!
Nicola

**Comment by martin-cs** [comment, marked as answer]
> Hi all,
> 
> I've been playing with CVC5 to do some verification of floating point
> code and it seems that it can handle mixed Real and FloatingPoint
> constraints very well.

That's great to hear!


> So I'm are wondering what is happening under the hood.

When terms of this kind are found (and registered) we produce a few
"obvious" lemmas:

https://github.com/cvc5/cvc5/blob/a8623b22f6f8d28191f090f8f5456540d9cb0a18/src/theory/fp/theory_fp.cpp#L553

These are things like `(fp.isZero f) => (= (fp.to_real f) 0)`.  These
handle the really obvious cases.  The rest are left until there is a
candidate model for both the floats and the reals.  Then every
conversion term we get the values in both parts of the model:

https://github.com/cvc5/cvc5/blob/a8623b22f6f8d28191f090f8f5456540d9cb0a18/src/theory/fp/theory_fp.cpp#L679

If they are equal then all is well and we have a consistent model.  If
not we lazily add some refinement lemmas:

https://github.com/cvc5/cvc5/blob/a8623b22f6f8d28191f090f8f5456540d9cb0a18/src/theory/fp/theory_fp.cpp#L173

Let's say we have `to_real(h)` so a term of Real sort but with
Floating-Point input. The real solver has `to_real(h) = 1/3` and the
float solver has `h = 0.25f`. (Note fractions for real literals,
decimal for float literals.)

If we just leave it to theory combination then we are likely going to
get one solver getting `!(to_real(h) = 1/3)` which ... doesn't narrow
the space down much.

Instead what we do is use the monotonicity of to_real and generate...

```
h >= 0.25f => to_real(h) >= 1/4
h <= 0.25f => to_real(h) <= 1/4
to_real(h) >= 1/3 => h >= 0.33333333f
to_real(h) <= 1/3 => h <= 0.33333332f
```


>  In particular, what are the theoretical guarantees for the theory
> combination of reals and floating points?

I haven't worked on this and to my knowledge no-one else has either. 
The implementation has at least one situation which can generate Zeno-
like non-termination.  This is in-principle fixable.


>  In principle, the combined quantifier-free fragment can be
> undecidable, so I'm wondering whether some well-known transfer result
> for satisfiability (such as Nelson-Oppen or Polite Combination) is
> employed. Floating Points do not seem to enjoy stable infiniteness,
> so I doubt that the Nelson-Oppen method can be applied.

We could, theoretically, instantiate all of these lemmas up front.  The
is gives a constant (for a fixed format) increase in formula size (the
constant is insane but ... ) and should ensure consistency between the
two theories.  I am not sure I see a problem here.

HTH

**Comment by nicola-gigante** [comment]
Thank you for the quick reply!

I was experimenting a bit more with unsatisfiable instances. Take a look at this problem:

```
(set-logic ALL)

(define-const onef Float32 ((_ to_fp 8 24) RNE 1))

(declare-const xf Float32)
(declare-const yf Float32)

(declare-const xr Real)
(declare-const yr Real)

(assert (= xf ((_ to_fp 8 24) RNE xr)))
(assert (= yf ((_ to_fp 8 24) RNE yr)))

(assert (>= xr 0))
(assert (<= xr 1))
(assert (>= yr 0))
(assert (<= yr 1))

(assert (> (+ xr yr) 1)) ; sat. with: (assert (< (+ xr yr) 1)) 
(assert (fp.lt (fp.add RNE xf yf) onef))

(check-sat)
```

The problem is unsatisfiable as it has two conflicting constraints (`xr + yr < 1` and `to_fp(xr) + to_fp(yr) > to_fp(1)`).
If you swap `<` for `>` in the second-last assert, it becomes satisfiable. 

The satisfiable version is solved instantly, while the unsatisfiable version makes CVC5 hang for minutes (I didn't wait for too long).

Do you have an idea of what's the problem here? Is this an inherent limitation of the current approach or do I just have to wait more for the answer?

Best,
Nicola

**Comment by martin-cs** [comment]
On Thu, 2022-02-17 at 07:27 -0800, Nicola Gigante wrote:
> 
> Thank you for the quick reply!
> 
> I was experimenting a bit more with unsatisfiable instances. Take a
> look at this problem:

<snip>

> The problem is unsatisfiable as it has two conflicting constraints
> (xr + yr < 1 and to_fp(xr) + to_fp(yr) > to_fp(1)).
> If you swap < for > in the second-last assert, it becomes
> satisfiable.
> 
> The satisfiable version is solved instantly, while the unsatisfiable
> version makes CVC5 hang for minutes (I didn't wait for too long).
> 
> Do you have an idea of what's the problem here? Is this an inherent
> limitation of the current approach or do I just have to wait more for
> the answer?

I suspect you are running into a limitation of the implementation (but
not of the approach).  If you consider the four inequalities that are
generated (which are, of course, not a Galois connection!) they are
sufficient but not the strongest thing you can give.  There are some
circumstances where you can get Zeno-like behaviour where the
constraints force the real value ever smaller but never actually cause
it to converge and thus give a constraint.  Continuing the above
example:

```
to_real(h) >= 1/3 => h >= 0.33333333f
to_real(h) <= 1/3 => h <= 0.33333333f
```

is correct but this is strictly more powerful:

```
to_real(h) >= lower => h >= 0.33333333f
to_real(h) <= upper => h <= 0.33333333f
```

where lower is the smallest real value such that:
`to_float(lower) = 0.33333333f`
and upper is the largest real value such that:
`to_float(upper) = 0.33333333f`

Instead of having a single real value where these fix the value of h,
we now have an interval which should stop the Zeno behaviour. 
Eventually.

I have implemented some of this strengthening already but not all of it
as it is quite fiddly (consider non-deterministic rounding-modes and
the edge behaviours around overflow and underflow).  I will be working
on this area again soon so if you really need this let me know.

Cheers,
 - Martin

**Reply by nicola-gigante** [reply]
Thanks again for the reply!

Yes, if it worked we would be able to use CVC5 for a verification technique for floating-point code we are working on. In particular, the solver works quite well on the small satisfiable instances we tried so far, but unsatisfiable instances are bite hard from the start. CVC5 is the only SMT solver that we found that supports reals and floating point together well enough to be useful for our purposes. We'd really appreciate any further progress on this front (although I realise that you may have different short-term priorities).

Thanks!
Nicola

**Reply by martin-cs** [reply]
No promises but I will see what I can do.

One thing that would help is if you could send me some examples to work on.  Or, better still, if you submitted them to SMT-COMP:

https://smt-comp.github.io/benchmark_submission.html

then we could have a QF_FPNRA or a better QF_FPLRA category and it would give all solvers the option to work on them!  I suspect COLIBRI and MathSAT could both do something interesting with them with a little work.

**Reply by nicola-gigante** [reply]
Sure, good idea! I'll let you know when the new benchmarks are ready.

Best,
Nicola

**Reply by nicola-gigante** [reply]
Hi!

I've translated to SMT some more examples.
The first simple one is this:
```
(set-logic ALL)

(declare-const x Float32)
(declare-const y Float32)

(declare-const rx Real)
(declare-const ry Real)

(assert (= x ((_ to_fp 8 24) RNE rx)))
(assert (= y ((_ to_fp 8 24) RNE ry)))

(assert (fp.gt (fp.mul RNE x y) ((_ to_fp 8 24) RNE 0)))
(assert (<= (* rx ry) 0))

(check-sat)
```

Here I test whether it is possible that a multiplication of any two floating point numbers results positive while the same multiplication in Real arithmetic results negative. This is obviously unsatisfiable but it is the archetype of the kind of tests that we want to do: can a given boolean expression made on top of floating point arithmetic give different results than the same expression computed in real arithmetic? In this simple example CVC5 returns correctly `unsat` very quickly.

The second example is more involved. The schema is the same but the computation is more complex. CVC5 hangs (not sure if it hangs or simply the solving time is too long), so it should be unsatisfiable as well.

```
(set-logic ALL)

(declare-const sx Float32)
(declare-const sy Float32)
(declare-const vx Float32)
(declare-const vy Float32)

(declare-const rsx Real)
(declare-const rsy Real)
(declare-const rvx Real)
(declare-const rvy Real)

(assert (= sx ((_ to_fp 8 24) RNE rsx)))
(assert (= sy ((_ to_fp 8 24) RNE rsy)))
(assert (= vx ((_ to_fp 8 24) RNE rvx)))
(assert (= vy ((_ to_fp 8 24) RNE rvy)))

(assert
  (fp.gt 
    (fp.mul RNE 
      (fp.add RNE
        (fp.mul RNE sx vx)
        (fp.mul RNE sy vy)
      )
      (fp.sub RNE
        (fp.mul RNE sx vx)
        (fp.mul RNE sy vy)
      )
    )
    ((_ to_fp 8 24) RNE 0)
  )
)

(assert
  (<=
    (* 
      (+
        (* rsx rvx)
        (* rsy rvy)
      )
      (-
        (* rsx rvx)
        (* rsy rvy)
      )
    )
    0
  )
)

(check-sat)
```

Hope it helps. When I'll have more examples I'll drop them here, and when I'll have more than a few I'll post them to the SMT competition.

Best,
Nicola

**Reply by martin-cs** [reply]
@barrettcw reminded me of this issue so perhaps an update and some new information.

My work on this has been delayed by poor health and rejected funding applications but I am still keen to improve this functionality.  There are two things that need to be done:

1. Generating the "overlapping" linking inequalities as described in https://github.com/cvc5/cvc5/discussions/8024#discussioncomment-2223945 .  This is 'just' an amount of engineering.  It will remove the non-termination issues from some UNSAT benchmarks.  It's a little bit hard to describe formally but cases where either the real or the float components of the expression have a small number of models should then always terminate in `O(min(#real models, #float models))`.  In fact it will be better than that ... it will be something like the number of hypercubes needed to partition the space covered by the projection of the models into each other theory but this is not so important.  The key thing is that if either the real or the float theory solver has enough information to reduce down to a small number of candidate models then it should terminate.

2. Unfortunately, this misses a big case of useful formula which are along the line of `real_expression - delta < float_expression < real_expression + delta`, i.e. bounding the error between a real computation and it's floating-point approximation.  These are harder because neither theory solver has enough information to do more than blindly suggest to the other what a model could be.  These will either formally not terminate or terminate after trying every possible set of floats, which is practical non-termination.  I have a paper design for solving these but it requires some small but very deep changes to the real solver, which, at the moment, we don't have anyone who is working on.

I hope cvc5 is still being useful for you and do keep me posted about your real/float successes and challenges.

**Reply by nicola-gigante** [reply]
HI @martin-cs, thank you for the update!

Yes, we are still interested in this kind of reasoning over floats with `cvc5`. As you said, the really useful formulas are those of the second kind. I don't have any experience with `cvc5`'s codebase but if I can help in some way let me know!
